src/main/java/VectorClockBackend.java

```diff
@@ -0,0 +1,157 @@
+                HashSet<Integer> FunctionIdList = ObjectTable.get(JavaObjectId).idlist;
+                TreeSet<GenericEffect> EffectList = ObjectTable.get(JavaObjectId).effectbuffer;
+                if (!FunctionIdList.contains(func.getId())) {
+                    FunctionIdList.add(func.getId());
+                    EffectList.add(updateEffect);
+                }
+            } catch (ClassCastException e) {
+                /* intentionally ignore this exception */
+                // This happens because we don't have an Idset for crdt initializations through
+                // updates so we got a redundant crdt object(something we already have)
+                assert (binary.getObject().getClass() == CRDT.class);
+            }
+        }
+
+        return JavaObjectId;
+    }
+
+    @Override
+    public OtpErlangBinary downstream(OtpErlangBinary JavaObjectId, OtpErlangBinary binary, OtpErlangMap clock) {
+        VectorClock effectClock = new VectorClock(clock);
+        GlobalClockTime.maxClock(effectClock);
+
+        OtpErlangBinary bin;
+        try {
+            CRDTEffect effect = new CRDTEffect((CRDT) binary.getObject(), effectClock);
+            bin = new OtpErlangBinary(effect);
+        } catch (ClassCastException e) {
+            GenericEffect effect = new GenericEffect((GenericFunction) binary.getObject(), effectClock);
+            bin = new OtpErlangBinary(effect);
+        }
+
+        return bin;
+    }
+
+    @Override
+    public OtpErlangTuple snapshot(OtpErlangBinary JavaObjectId) throws NoSuchObjectException {
+        // assert that object is in table else request it
+        CRDT crdt_object = ObjectTable.get(JavaObjectId).object;
+        if (crdt_object == null) {
+            throw new NoSuchObjectException();
+        }
+        CRDT new_crdt_object = crdt_object.deepClone();
+        OtpErlangBinary new_key = newJavaObjectId();
+
+        // ObjectTable.get(JavaObjectId).idlist.clear();
+        // ObjectTable.remove(ERLObjectId);
+
+        // Apply any effects that have passed the clock time to the new copy of the
+        // object
+        TreeSet<GenericEffect> new_crdt_effect_buffer = new TreeSet<GenericEffect>();
+        for (GenericEffect e : ObjectTable.get(JavaObjectId).effectbuffer) {
+            // Use a correct compare based on types
+            if (e.time.lessthan(this.GlobalClockTime)) {
+                new_crdt_object.invoke((GenericFunction) e.func);
+            } else {
+                new_crdt_effect_buffer.add(e);
+            }
+        }
+
+        CRDTMapEntry maptriple = new CRDTMapEntry(new_crdt_object, new HashSet<Integer>(), new_crdt_effect_buffer);
+        ObjectTable.put(new_key, maptriple);
+        OtpErlangObject[] emptypayload = new OtpErlangObject[2];
+        emptypayload[0] = new_key;
+        emptypayload[1] = new OtpErlangBinary(crdt_object);
+        OtpErlangTuple new_snapshot = new OtpErlangTuple(emptypayload);
+        return new_snapshot;
+    }
```

Comment:
Your snapshots must contain all updates which are more recent than the "global min time"

Reply:
Whoops